htlcswitch/htlcnotifier.go

```diff
@@ -0,0 +1,276 @@
+package htlcswitch
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/subscribe"
+)
+
+// HtlcNotifier notifies clients of htlc forwards, failures and settles for
+// htlcs that the switch handles.  It takes subscriptions for its events and
+// notifies them when htlc events occur. These are served on a best-effort
+// basis; events are not persisted, delivery is not guaranteed (in the event
+// of a crash in the switch, forward events may be lost) and events will be
+// replayed upon restart. Events consumed from this package should be
+// de-duplicated, and not relied upon for critical operations.
+//
+// The htlc notifier sends the following kinds of events:
+// Forwarding Event:
+// - Represents a htlc which is forwarded onward form our node.
+// - Present for htlc forwards through our node and local sends.
+//
+// Link Failure Event:
+// - Indicates that a htlc has failed on our incoming or outgoing link,
+//   with an incoming boolean which indicates where the failure occurred.
+// - Incoming link failures are present for failed attempts to pay one of
+//   our invoices (insufficient amount or mpp timeout, for example) and for
+//   forwards that we cannot decode to forward onwards.
+// - Outgoing link failures are present for forwards or local payments that
+//   do not meet our outgoing link's policy (insufficient fees, for example)
+//   and when we fail to forward the payment on (insufficient outgoing
+//   capacity, or an unknown outgoing link).
+//
+// Forwarding Failure Event:
+// - Forwarding failures indicate that a htlc we forwarded has failed at
+//   another node down the route.
+// - Present for local sends and htlc forwards which fail after they left
+//   our node.
+//
+// Settle event:
+// - Settle events are present when a htlc which we added is settled through
+//   the release of a preimage.
+// - Present for local receives, and successful local sends or forwards.
+//
+// Each htlc is identified with a unique htlc key which contains its incoming
+// and outgoing channel id and index. Htlcs, and their subsequent settles or
+// fails, can be identified by these four elements. Note that sends from our
+// node will have a zero incoming channel id/index, because they originated
+// locally, and receives will have a zero outgoing channel id/index because
+// they terminate at our node.
+type HtlcNotifier struct {
+	started sync.Once
+	stopped sync.Once
+
+	ntfnServer *subscribe.Server
```

Comment:
Can `subscribe.Server` be embedded, so that the `Start/Stop functions are inherited?

Reply:
Like, resolved, but no change? :)

Reply:
ah sorry, was conflating it with [this](https://github.com/lightningnetwork/lnd/pull/3781#discussion_r378064851) so resolved to de-dup. 

I like having the log line for starting like we have for other notifiers (eg `ChannelNotifier`).

Reply:
Ok